### PR TITLE
Support app level API tokens

### PR DESCRIPTION
**Pull request by hagould**
These changes are updating the tokens create, tokens list and tokens delete command to account for new arg parameter --type which allows you to choose to create an app API token when an app context or app command parameter is present.
These commands will default to user type tokens unless app is specifically requested.

In order to make and test these changes this PR regenerated the API clients with the app token APIs manually edited to be exposed for generation - in reality they're currently x-internal. 

Once the app API token work is validated and the routes are updated, the API client commit will need to be reverted and regenerated.

`npm run download-swagger` or `npm run autorest`,
then `npm run autorest:gen` 

This PR is dependent on https://github.com/microsoft/appcenter-cli/pull/852
This branch has been merged with master and is ready to go

### CODE REVIEW

**Comment by jp-andre** [comment]
Can we separate the regen of API from the code changes? I think we updated the client APIs recently in master branch, maybe you should be able to simply merge master and work from that?

File: src/commands/tokens/create.ts [hunk 1]
```diff
@@ -28,15 +31,37 @@ export default class TokenCreateCommand extends Command {
   @hasArg
   description: string;
 
+  @help("The type of token principal authentication: [" + allPrincipalTypes.join(", ") + "]")
+  @shortName("t")
+  @longName("type")
+  @hasArg
+  @defaultValue("user")
+  public principalType: string;
```
Comment (jp-andre) [inline src/commands/tokens/create.ts:39]: Any reason to make this `public`?

File: src/commands/tokens/delete.ts [hunk 2]
```diff
@@ -27,22 +33,42 @@ export default class AppDeleteCommand extends Command {
...
+  @help("The type of token principal authentication: [" + allPrincipalTypes.join(", ") + "]")
+  @shortName("t")
+  @longName("type")
+  @hasArg
+  @defaultValue("user")
+  public principalType: string;
+
   async run(client: AppCenterClient): Promise<CommandResult> {
-    const confirmation = await prompt.confirm(`Do you really want to delete the token with ID "${this.id}"`);
+    const tokenLevel = this.principalType === PrincipalType.USER ? principalMessaging.user : principalMessaging.app;
```
Comment (jp-andre) [inline src/commands/tokens/delete.ts:44]: Can we have a validation function that verifies that an acceptable option is passed? We want to have exactly `user` or `app`, or unspecified (default).
Some commands have `validateParameters()`.

File: src/util/misc/principal-type.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export enum PrincipalType {
+  USER = "user",
+  APP = "app",
+}
+
+export const allPrincipalTypes = [PrincipalType.USER, PrincipalType.APP];
+
+export const principalMessaging = {
+  [PrincipalType.USER]: "user type",
+  [PrincipalType.APP]: "app type",
```
Comment (jp-andre) [inline src/util/misc/principal-type.ts:10]: Do we really need this? It seems that simply sticking to PrincipalType value is enough.

File: test/commands/tokens/create-test.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
+import { expect, use } from "chai";
```
Comment (jp-andre) [inline test/commands/tokens/create-test.ts:1]: Thanks so much for creating test cases here! 🎉

File: test/commands/tokens/create-test.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      id: fakeId,
+      description: fakeDescription,
+      api_token: "token",
+      scope: "all",
+    });
+
+    const command = new CreateTokenCommand(getCommandArgs([]));
+    const result = await command.execute();
+
+    expect(result.succeeded).to.be.true;
```
Comment (jp-andre) [inline test/commands/tokens/create-test.ts:88]: Maybe also check somehow that type is user? Is that possible?

**Comment by jp-andre** [comment]
As our team will be on vacation from Apr 30 to May 5 included, let's wait until full review after May 5 to merge this. Also we'll be expecting a separate PR for the API update.

File: src/util/misc/principal-type.ts (hunk 3 above)
Comment (zacharysierakowski) [reply src/util/misc/principal-type.ts:10]: Agreed, I was going to say the same!

**Comment by zacharysierakowski** [comment]
> Can we separate the regen of API from the code changes? I think we updated the client APIs recently in master branch, maybe you should be able to simply merge master and work from that?

I think we will do this all in one PR actually, if that's okay? We plan to update the PR either on Friday or Monday (May 4th) with the "correct" regen.

File: src/commands/tokens/create.ts (hunk 1 above)
Comment (hagould) [reply src/commands/tokens/create.ts:39]: This formatting matched other CLI command arguments (I was looking at the distribute ones for reference), but no I didn't actually realize that the previous parameters for these commands weren't marked as public. Would you prefer without `public`?

File: test/commands/tokens/create-test.ts (hunk 5 above)
Comment (hagould) [reply test/commands/tokens/create-test.ts:88]: the test only mocks that the user endpoint is being called - which validates that the command defaults to user type token. I might be able to mock the app token endpoint and make sure that it's not being called, but it shouldn't be necessary to cover this case

**Comment by jp-andre** [comment]
Looks like the build failed, let's verify again after the merge conflicts have been resolved.

File: src/commands/tokens/delete.ts [hunk 6]
```diff
@@ -27,22 +33,40 @@ export default class AppDeleteCommand extends Command {
...
+      );
+    } else if (this.principalType === PrincipalType.APP) {
+      const app: DefaultApp = this.app;
+      deleteTokenResponse = await out.progress(
+        tokenMessaging,
+        clientRequest<null>((cb) => client.appApiTokens.deleteMethod(app.ownerName, app.appName, this.id, cb))
+      );
     }
 
+    if (deleteTokenResponse.response.statusCode === 404) {
```
[the remark this answers is not in the document]
Comment (hagould) [reply src/commands/tokens/delete.ts:67]: My original intent was to not overly change the behavior of these commands when they are already are being used for user tokens, but I think this shouldn't realistically cause problems

**Comment by hagould** [comment]
the component governance in the build tasks is catching an issue with yarg-parser in the dev gulp dependency. Anyone seen this before or aware of what's best to do here? I don't believe it's something that can be fixed with an audit

**Pull request merged**